### ISSUE TITLE
SpotBugs should warn about file descriptor leak for Files.list().forEach() unclosed stream

### ISSUE THREAD

**Issue by jamesdaily** [spotbugs/spotbugs#650]
SpotBugs should warn about a potential file descriptor leak when invoking Files.list() without later closing the stream.

Please refer to the following gist: https://gist.github.com/jamesdaily/2f938b52235b400089d96e8ea248b8ae

Creating the Stream using Try-with-resources is auto-closed and should not warn.

Creating the Stream using a regular Try block (omitting `finally {stream.close();}`), the compiler does helpfully warn: `Resource leak: 'files' is never closed` , but SpotBugs does not also warn. Should it?

Creating the Stream in a regular Try block and then chaining .forEach(), as may be a popular way to code, the compiler does NOT warn, and neither does SpotBugs, but perhaps they should, as .forEach() does not close the Stream.

**Comment by KengoTODA** [comment]
[Stream#forEach()](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-) is a [terminal operation](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#StreamOps), so we can treat that this stream is already consumed [and closed](https://github.com/dmlloyd/openjdk/blob/jdk8u/jdk8u/jdk/src/share/classes/java/nio/file/Files.java#L3476).

And in your sample code, I can no opened file. `Files.list()` returns just `Path` and it won't open any file. Which `file` do you mean that you need to close explicitly?

**Comment by jamesdaily** [comment]
Hello, per the Files.list() javadoc:

> The returned stream encapsulates a {@link DirectoryStream}. If timely disposal of file system resources is required, the {@code try}-with-resources construct should be used to ensure that the stream's {@link Stream#close close} method is invoked after the stream operations are completed.

As for which file is open, the folder itself is left open and after many invocations, perhaps called on a frequent scheduler, will eventually exceed the max allowed open files (on Linux, anyway).

**Comment by jamesdaily** [comment]
Does this still need info? I hope my response answered your questions.

**Comment by KengoTODA** [comment]
I still think that the calling terminal operation should be enough. It is defined to close stream certainly.

I can understand the following case should be problem:

```java
Stream<Path> lists = Files.list(path);
// do something... then we may keep 'lists' open even after some exceptions are thrown
lists.forEach(...);
```

but if we code `Files.list(path).forEach(...)` then I'm sure that the Stream is closed certainly.

**Comment by KengoTODA** [comment]
I found that my understanding is incorrect. By following code, we can confirm that terminal operation won't close the created stream instance.

```java
        Stream<Path> stream = Files.list(Paths.get("."));
        stream.onClose(() -> {
            System.out.println("closed");
        });
        stream.forEach(path -> {
            throw new RuntimeException("");
        });
```

**Comment by jamesdaily** [comment]
Thanks, @KengoTODA. Do you think the "need info" tag can be removed at this time?

**Comment by bitzl** [comment]
I just discovered this issue by looking for a check like this. We'd be happy if Spotbugs would detect the use of `Files.list()` without  a try-with-statement. Ist there anything we can do to help?
